File: docs/basic/code/greedy/greedy_1.py
from collections import defaultdict
from heapq import heappush, heappop

a = defaultdict(list)
for _ in range(int(input())):
    d, p = map(int, input().split())
    a[d].append(p)  # Lưu lợi nhuận ứng với thời hạn này.

ans = 0  # Ghi lại tổng lợi nhuận.
q = []  # Min-heap duy trì giá trị nhỏ nhất.
l = sorted(a.keys(), reverse=True)
for i, j in zip(l, l[1:] + [0]):
    for k in a.pop(i):
        heappush(q, ~k)
    for _ in range(i - j):
        if q:  # Lấy công việc có lợi nhuận lớn nhất khỏi heap.
            ans += ~heappop(q)
        else:  # Thoát vòng lặp khi heap rỗng.
            break
print(ans)

File: docs/basic/code/greedy/greedy_1.cpp
#include <algorithm>
#include <cmath>
#include <cstring>
#include <iostream>
#include <queue>
using namespace std;

struct f {
  long long d;
  long long p;
} a[100005];

bool cmp(f A, f B) { return A.d < B.d; }

// Min-heap duy trì giá trị nhỏ nhất.
priority_queue<long long, vector<long long>, greater<long long>> q;

int main() {
  long long n, i;
  cin >> n;
  for (i = 1; i <= n; i++) {
    cin >> a[i].d >> a[i].p;
  }
  sort(a + 1, a + n + 1, cmp);
  long long ans = 0;
  for (i = 1; i <= n; i++) {
    if (a[i].d <= (int)q.size()) {  // Vượt quá hạn chót.
      if (q.top() < a[i].p) {       // Đổi lựa chọn trước đó.
        ans += a[i].p - q.top();
        q.pop();
        q.push(a[i].p);
      }
    } else {  // Đưa trực tiếp vào hàng đợi.
      ans += a[i].p;
      q.push(a[i].p);
    }
  }
  cout << ans << endl;
  return 0;
}
